pkg/platform/runtime/buildplan/dependencies.go

```diff
@@ -0,0 +1,192 @@
+		return nil, nil
+	}
+
+	currentArtifact, ok := lookup[artifactID].(*model.Artifact)
+	if !ok {
+		return nil, errs.New("Incorrect target type for id %s, expected Artifact", artifactID)
+	}
+
+	// We iterate through the direct dependencies of the artifact
+	// and recursively add all of the dependencies of those artifacts map.
+	// This is the same as the runtime closure calculation.
+	for _, depID := range currentArtifact.RuntimeDependencies {
+		deps[depID] = struct{}{}
+		artifactDeps := make(map[strfmt.UUID]struct{})
+		_, err := buildBuildClosureDependencies(depID, lookup, artifactDeps, result)
+		if err != nil {
+			return nil, errs.Wrap(err, "Could not build map for runtime dependencies of artifact %s", currentArtifact.NodeID)
+		}
+	}
+
+	// Here we iterate through the inputs of the step that generated the
+	// artifact, specifically the inputs that are tagged as dependencies.
+	// We recursively add all of the dependencies of the step intputs to
+	// the result map. This is the buildtime closure calculation.
+	step, ok := lookup[currentArtifact.GeneratedBy].(*model.Step)
+	if !ok {
+		// Artifact was not generated by a step, skipping because these
+		// artifacts do not need to be built.
+		return nil, nil
+	}
+
+	// We iterate through the inputs of the step that generated the
+	// artifact, specifically the inputs that are tagged as dependencies and
+	// build a build-time closure for each.
+	for _, input := range step.Inputs {
+		if input.Tag != model.TagDependency {
+			continue
+		}
+
+		for _, inputID := range input.NodeIDs {
+			deps[inputID] = struct{}{}
+			_, err := buildBuildClosureDependencies(inputID, lookup, deps, result)
+			if err != nil {
+				return nil, errs.Wrap(err, "Could not build map for step dependencies of artifact %s", currentArtifact.NodeID)
+			}
+		}
+	}
+
+	var uniqueDeps []strfmt.UUID
+	for id := range deps {
+		uniqueDeps = append(uniqueDeps, id)
+	}
+
+	info, err := getSourceInfo(currentArtifact.GeneratedBy, lookup)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not resolve source information")
+	}
+
+	result[strfmt.UUID(currentArtifact.NodeID)] = artifact.Artifact{
+		ArtifactID:       strfmt.UUID(currentArtifact.NodeID),
+		Name:             info.Name,
+		Namespace:        info.Namespace,
+		Version:          &info.Version,
+		RequestedByOrder: true,
+		GeneratedBy:      currentArtifact.GeneratedBy,
+		Dependencies:     uniqueDeps,
+		URL:              currentArtifact.URL,
+	}
+
+	return deps, nil
+}
+
+// DependencyTreeFor returns for the given artifact ID a map of dependencies to a list of unique
+// subdependencies. The returned map excludes any dependencies in the given filter, unless there is
+// a version mismatch and `keepUpdated` is true (this signifies a dependency's version was
+// updated).
+func DependencyTreeFor(a artifact.ArtifactID, artifacts artifact.Map, filter artifact.Map, keepUpdated bool) map[artifact.ArtifactID][]artifact.ArtifactID {
```

Comment:
Nit: Calling this a tree feels misleading since the second level is essentially flattening the rest of the tree.

Perhaps for future use this could actually construct the whole tree, but on each level give a count summary of what's underneath. 

Anyway, strong nit; happy to accept as is.

Reply:
I've renamed the function. If you still don't like it, please suggest a better one :)